File: Day 4/Covariance.java
class Animal {}
class Cat extends Animal {}
class Dog extends Animal {}

public class Covariance {
    public static void main(String[] args) {
        // Java arrays are covariant:
        Cat[] cats = new Cat[2];
        
        // Allowed in compile time 
        Animal[] animals = cats;

        try {
            // RUNTIME ERROR: ArrayStoreException
            animals[0] = new Dog();   
        } catch (Exception ex) {
            System.out.println("Runtime Safety Check Failed" + ex);
        }
    }
}

File: Day 4/Covariance.ts
class Animal { }
class Cat extends Animal { meow() {} }
class Dog extends Animal { bark() {} }

function main() {
    const cats: Cat[] = [new Cat()];

    // arrays are covariant in TypeScript.
    const animals: Animal[] = cats;

    // UNSAFE: inserting a Dog into a Cat[] via Animal[]
    animals.push(new Dog());

    
    // cats now contains a Dog in position 1.
    console.log("cats array:", cats);

    
    cats[1].meow();
    // It will crash: meow() does not exist on Dog.
}
main();
